### ISSUE TITLE
Raster file type

### ISSUE THREAD

**Issue by pkdash** [hydroshare/hydroshare#1477]
Relates to issue #1034 

This is the first file type to be implemented. Requirements for raster file type will be documented in this issue.

TODO:
- [x] Implement raster file type to help design data model
       - When a content file (tif or zip) is set to Raster File Type system will create a folder and put all raster related files in that folder.
- [x] Implement GenericFile Type 
       - Any file uploaded to Composite resource by default will be assigned GenericFile type
       - GenericFile type will have only 'Coverage" and 'Key/Value' metadata. Both are optional. 
- [ ] Update terms page to include file type metadata terms and composite resource type information

Requirements/Assumptions:
 - The raster file type metadata consists of same metadata as that of the raster resource specific metadata
- Supported file mime types same as raster resource type
- Uploaded file mime types same as raster resource type
- Logical file type 'GeoRasterLogicalFile'
- File operation business rules same as that of raster resource content files. Files that are part of GeoRaster file type can't be renamed. The folder that represents the GeoRaster file type can be renamed and deleted. The same folder can be zipped.
- System will not attempt to extract metadata on file upload to Composite Resource (at least in the first phase of implementation) 
- Composite resource "Coverage' metadata can't be edited by user. The "Coverage" metadata element will be updated by the system based on the "Coverage" metadata of files in the resource.
- There will be an option (Set to Geo Raster File Type) to extract metadata from any tif/zip file in a composite resource.
  - If the selected file is a valid raster file, then metadata will be extracted and stored in association with the uploaded file's logical file type instance (GeoRasterLogicalFile). Also a folder will be created to store one or more tif files and a vrt file.
  - User will not be allowed to put any other files to the folder that is of type GeoRaster.
  - If the user deletes a file from the GeoRaster folder, all files in that folder will be deleted including the folder. UI doesn't allow deleting a file that is part of the geo raster file type.
  - User can edit metadata associated with the GeoRaster logical file.
  - Metadata associated with a file will be written to the resourcemetadata.xml file.

Requirements related to 'Coverage' metadata:
   - The composite resource Coverage data (both spatial and temporal) will not be user editable. These coverage data will be automatically updated by the system based on the coverage data that exists for each of the logical file objects (Generic and Geo Raster) in the resource. The composite resource coverage data will be superset of coverage data in all of the logical file objects in the resource.
   - For raster logical file type, the coverage metadata will be extracted by the system when the user set the a tif or zip file to raster file type. Use will not be able to edit the extracted coverage data. Both the converted coverage data as well as the original extracted coverage (spatial reference or original coverage) will be stored at the logical file object. In addition, user will be able to enter temporal metadata for raster file type (optional).
   -  For generic logical file, the user can enter spatial and temporal coverage data.
   - Coverage projection will be always in WGS 84.
   - Coverage data is optional for generic logical file.
   - Coverage data is required for raster logical file.

**Comment by pkdash** [comment]
@dtarb and @gantian127  Since I am implementing raster file type within a new resource type - Composite Resource, I am assuming that each raster file type needs to have the required metadata elements before a resource can be made public.

**Comment by dtarb** [comment]
Yes, I agree.

**Comment by pkdash** [comment]
@dtarb Is there any restriction what type of coverage be allowed for Generic File type? At the resource level we have 2 types of coverage - temporal and spatial (box or point).

**Comment by dtarb** [comment]
There is a simple option and a general option.  I am OK with going with the simple option first but would like @gantian127 to scope out and specify requirements for the general option which would handle generic file types the same as raster and feature file types.

The simple option is that it should be exactly the same as what is currently used for Generic resources.  This is a point or a box that has to have units=Decimal degrees and projection=WGS 84 EPSG:4326 that would have metadata like

```
<dc:coverage>
<dcterms:box>
<rdf:value>northlimit=42.2173044595; eastlimit=-111.527984414; southlimit=41.9934109403; westlimit=-111.857693756; units=Decimal degrees; projection=WGS 84 EPSG:4326</rdf:value>
</dcterms:box>
</dc:coverage>
```
or

```
<dc:coverage>
<dcterms:point>
<rdf:value>east=-111.5778656; north=41.8285177581; units=Decimal degrees; projection=WGS 84 EPSG:4326</rdf:value>
</dcterms:point>
</dc:coverage>
```

The general option would allow the logical file object coverage to have any projection and units and would require projection and unit conversions in determining the resource level coverage that is always represented in WGS-84 EPSG:4236 coordinate system.  Its coverage element could be modeled after the spatial reference elements of geographic raster and geographic feature resources and would need to include
- point or box defined using (east, north) or (eastlimit, southlimit, northlimit, westlimit)
- projection_name
- EPSG code (or other authority code)
- units
- datum
- projection string (WKID string)

**Comment by pkdash** [comment]
@dtarb and @gantian127 I have documented the requirements related to 'Coverage' metadata at the beginning of this issue based on the discussion we had this morning. If I have missed anything, feel free to edit.

**Comment by gantian127** [comment]
@pkdash thank you for the update.

**Comment by pkdash** [comment]
@hyi and @alvacouch 
The current strategy of regenerating the xml metadata files every time we update any of the metadata elements or do any CRUD operation on resource content files is going to be a big problem with the Composite Resource type in which all different file types have their own metadata. This is the same problem that Alva recently discovered as it relates to updating individual metadata elements though REST call.

To solve this problem I am planning to make the following changes:

1.	Set resource AVU flag  ‘modified’ to ‘true’ each time metadata elements get updated or any content file operation.
2.	Regenerate xml metadata files as part of bag creation function. The bag creation function will call the create_bag_files().
3.	If the user tries to download either the ‘resourcemap.xml’ or ‘resourcemetadata.xml’ the download function will check for resource ‘modified’ AVU flag, and if it is found to be ‘true’ then run the create_bag_files() function prior to the download of the file.

Let me know if you see any issues with the above strategy.

**Comment by hyi** [comment]
@pkdash The changes you outlined above to be made sounds good and should work.

**Comment by alvacouch** [comment]
@pkdash This approach solves bagging but does not solve the use case where someone updates metadata and then downloads sciencemetadata.xml expecting changes to be present ... we also have to regenerate the file before any scimeta file download.  It will be common for apps to avoid bagging and download individual files. However, at present, I plan to satisfy individual element queries direct from django so they won't be affected.

**Comment by pkdash** [comment]
@alvacouch See item#3.

**Comment by alvacouch** [comment]
That should work...

**Comment by pkdash** [comment]
This has been implemented and deployed as part of 1.9.4.
